preeti13456/Blockchain-project
# Blind Signature Scheme Implemented

# Digital Signature Verification with Blind Signature Scheme Support - This Update Addresses Issue #7

## **Blind Signature with Elliptic Curve Digital Signature Algorithm feature** has been implemented successfully 

### Feature Description
The feature ensures that the voters can get their ballot (vote) signed by election authority before casting it. It also ensures that the signatures are verified and not tampered by any third party. This makes the voting process highly secure, authorized and hassle-free.
Also, a critical bug related to web3 library modules inclusion has been fixed.

### How the Project Implements Signing & Verification 
- When Voter submits the vote from the UI, his vote is then **hashed(blinded)** and signed with **Election Authority's private key** on the front end using **Web3 JS** (the *public key counterpart* is initially set during the contract deployment)
- According to Blind signature scheme, Election authority DO NOT have access to the message's content and can only sign and send it back (Implemented in _signBlindVote()_ function) to it’s caller.
- Voter's **vote** along with **blind signature** is then sent to the blockchain contract 
- Over the contract, the message's state before it was signed is regenerated by performing a number of steps.
    - abi-encoding and packing the parameters
    - hashing it with **Keccak256 (SHA-3)** algorithm 
    - Adding Ethereum's predefined PREFIX + MESSAGE_LENGTH + MESSAGE
    - hashing the output again with **Keccak256**

- Then the **ECC parameters (r,s,v)** are extracted from the received **signature**
- Solidity's _ecrecover()_ function then applies the core mathematics of **ECC (Elliptic Curve Cryptography)**  on the calculated hash and the received signature to generate the **signer's public key (Election organizer)** and then it is compared with the stored public key
- If these two matches, the signature is verified and the vote along with the blind signature is recorded on the blockchain
- I have performed the entire scheme using only **Web3 JS**, without the inclusion of any third party library such as **ethereumjs-abi, OpenZeppelin** etc.

### Bug-fix
On the front end, the function needed to hash (Keccak256) the message was provided in Web3 JS library (**web3.utils.soliditySha3()**), but it was not being recognised as the *web3-utils* sub-package is now distributed as separate modules. The front end’s web3 was loaded with an old version of web3.min.js on the *src/js* directory, so I used **CDN to fetch the latest web3.min.js** and the hashing function worked fine.

## Types of change


## How Has This Been Tested?

1. Fire up **Ganache**, take a demo signer's address (assume that as *Election Authority's Account*) and place it on *signerAddress* variable on *election.sol* contract.
![qwe](https://user-images.githubusercontent.com/42487891/90331909-1f58cf80-dfda-11ea-969c-33385e3c8e9f.png)

2. Run `$ truffle migrate --reset` to migrate the contracts as fresh ones.
3. Copy your deployed contract address and take the demo **Election Authority's private key** to paste them on the *contractAddress* and *signerAccount* variables accordingly on *app.js* script.
It  would display -
![origAddress](https://user-images.githubusercontent.com/42487891/90313986-422fa900-df32-11ea-9bec-7d9b7dcfb955.png)

4. Run `'$ npm run dev'`, open browser with **MetaMask** activated and connect with an account with sufficient Ethers (make sure that is not the signer’s account).

5. Register as a voter with *NID*, wait for transaction to be mined, and then refresh the page.
Click on *Vote*. Under the hood, the vote will be hashed, signed and sent over to the blockchain just fine. (You can check the broadcasted events on the browser console)

6. Now switch account, register as another voter and tamper the original signer address on *app.js* script by changing some random characters.
![tampered Address](https://user-images.githubusercontent.com/42487891/90314016-7c994600-df32-11ea-918c-85b8d7cf1497.png)

7. Now try to vote again and you will be notified of warning from the contract side, so you can’t won’t be able to make the successful transaction even if you try to sign the vote with any other private key except for the organizer’s one.


![transaction error](https://user-images.githubusercontent.com/42487891/90314029-9175d980-df32-11ea-821d-81ebead083b5.png)
### This ensures that the ballots are guaranteed to be authorized.
## Successful Tests
- [x] `Initializes with two voters`
- [x] `initializes with two candidates`
- [x] `it initializes the candidates with the correct values`
- [x] `Initializes the voters with the correct values`
- [ ] `Successful vote casting by single voter` (requires fixing after update)
- [ ] `Invalid candidate voting exception` (requires fixing after update)
- [ ] `Double voting exception` (requires fixing after update)

## Test Configuration:
* Firmware version: Win 10 64 bit
* Hardware: Intel Core i5 4460
* Truffle : v5.1.10 (core: 5.1.10)
* Solidity - 0.4.23 (solc-js)
* Node - v12.18.2
* Web3.js - v1.2.1


## Checklist:
